### ISSUE TITLE
Improve rule definitions for PMD: AvoidCalendarDateCreation, AvoidConcatInAppend and AvoidConcatInLoop

### ISSUE THREAD

**Issue by jborgers** [jborgers/PMD-jPinpoint-rules#26]
A.o. add examples, standardize attribute usage, correct link.
Investigate: 
For the concat rules, they don't seem to consider string literals in a special way… when writing `"foo" + "bar"`, the compiler will do the concat in compile time, so there is no real performance penalty (same happens for final strings initialized upon declaration, but that one is way harder to detect under the current XPath support; and I'm unsure, but probably Java 8+ does the same for effectively final variables).

**Comment by jborgers** [comment]
I looked into the concatenation with (effectively) finals. 

I assumed concatenation of literals and static final fields is optimized by the javac compiler. And I read about Indifying String concat in Java 9. 
First I found that javac can optimize also for final locals with constant expression, see: 
https://stackoverflow.com/questions/49868917/javac-missing-optimization-for-effective-final
So, based on this, we should assume local final fields are constant expressions, so have:
private String testFinalLocalGood() {
    final String local = "local"; // final assumed a constant expression
    return new StringBuilder().append("fld:" + local).toString(); // good
}
To avoid false positives. Checking if local is a constant expression would make the rule too complicated. So I just assume final locals are good. 

If it is not a final, concatenation is in 9+ deferred to runtime and it can do optimizations like not use boxing for primitives, optimize buffer length.
Indifying resources:
https://shipilev.net/talks/jfokus-Feb2016-lord-of-the-strings.pdf
https://openjdk.java.net/jeps/280
No easy reads. I assume:
private String testNonFinalLocal() {
    String local = "local"; // effectively final not as fast as final constant exp
    return new StringBuilder().append("fld:" + local).toString();
} 
Will be faster with 9+ than before, however, I assume still faster is:
private String testNonFinalLocalGood() {
    String local = "local"; 
    return new StringBuilder().append("fld:").append(local).toString();
} 
We should actually benchmark with JMH to be sure.
